In [19]:
import pandas as pd
import pickle

with open("./raw_data/allActs.pickle", "rb") as file:
    data = pickle.load(file)

df = pd.json_normalize(data)
df

,act.type,act.value,title.type,title.value,publicationDate.type,publicationDate.value,entryintoforce.type,entryintoforce.value,typeDoc.type,typeDoc.xml:lang,typeDoc.value,fileUrl.type,fileUrl.value
0,uri,https://fedlex.data.admin.ch/eli/oc/1/82_69_80,literal,Trattato d'estradizione tra la Svizzera e l'Im...,literal,1874-06-03,literal,1874-06-03,literal,it,Trattato internazionale bilaterale,NaN,NaN
1,uri,https://fedlex.data.admin.ch/eli/oc/18/654_588...,literal,Ordinanza concernente la consegna e il control...,literal,1901-04-19,literal,1901-04-19,literal,it,Ordinanza del Consiglio federale,NaN,NaN
2,uri,https://fedlex.data.admin.ch/eli/oc/18/865_796...,literal,Ordinanza del 9 novembre 1901 sull'esercizio d...,literal,1901-11-09,literal,1901-11-09,literal,it,Ordinanza del Consiglio federale,NaN,NaN
3,uri,https://fedlex.data.admin.ch/eli/oc/18/870_801...,literal,Ordinanza sull'esercizio della Fabbrica federa...,literal,1901-11-09,literal,1901-11-09,literal,it,Ordinanza del Consiglio federale,NaN,NaN
4,uri,https://fedlex.data.admin.ch/eli/oc/19/816_779...,literal,Decreto del Consiglio federale del 24 dicembre...,literal,1906-07-01,literal,1906-07-01,literal,it,Ordinanza del Consiglio federale,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
46850,uri,https://fedlex.data.admin.ch/eli/cc/2024/142,literal,Ordinanza della FINMA del 6 marzo 2024 sui ris...,literal,2024-04-05,literal,2025-01-01,literal,it,Atto legislativo delle aziende e degli stabili...,NaN,NaN
46851,uri,https://fedlex.data.admin.ch/eli/cc/2024/331,literal,Legge federale del 30 settembre 2022 sulla pro...,literal,2024-07-02,literal,2025-01-01,literal,it,Legge federale,NaN,NaN
46852,uri,https://fedlex.data.admin.ch/eli/cc/2024/493,literal,Ordinanza del 28 agosto 2024 sull’adeguamento ...,literal,2024-09-13,literal,2025-01-01,literal,it,Ordinanza del Consiglio federale,NaN,NaN
46853,uri,https://fedlex.data.admin.ch/eli/cc/2024/685,literal,Ordinanza del 13 novembre 2024 sulla cooperazi...,literal,2024-11-25,literal,2025-01-01,literal,it,Ordinanza del Consiglio federale,NaN,NaN


In [ ]:
# data filtered by keeping only the rows with a corresponding xml file
df_xml = df[df['fileUrl.value'].notna()]
df_xml = df_xml[df_xml['fileUrl.value'].str.contains('.xml')]

# display the number of laws per publication year
df_countYear = df_xml
df_countYear['publicationDate.value'] = df_countYear['publicationDate.value'].str[:4]
print(df_countYear['publicationDate.value'].value_counts().sort_index(ascending=False))

print("------------------------------------------------------------------------------------")

# display the number of laws per type of document
df_countType = df_xml
df_countType['typeDoc.value'] = df_countType['typeDoc.value']
print(df_countType['typeDoc.value'].value_counts())

publicationDate.value
2025      8
2024    844
2023    870
2022    571
Name: count, dtype: int64
------------------------------------------------------------------------------------
typeDoc.value
Ordinanza del Consiglio federale                                                 575
Ordinanza di un dipartimento                                                     335
Ordinanza di un'ufficio                                                          320
Campo d'applicazione                                                             297
Trattato internazionale bilaterale                                               186
Correzione RU                                                                    122
Trattato internazionale multilaterale                                            100
Legge federale                                                                    93
Atto legislativo delle aziende e degli stabilimenti autonomi                      82
Decreto del Consiglio federale          

,act.type,act.value,title.type,title.value,publicationDate.type,publicationDate.value,entryintoforce.type,entryintoforce.value,typeDoc.type,typeDoc.xml:lang,typeDoc.value,fileUrl.type,fileUrl.value
25122,uri,https://fedlex.data.admin.ch/eli/oc/2023/595,literal,Comunicazione concernente dei trattati interna...,literal,2023,literal,1997-12-03,literal,it,Comunicazione di abrogazioni (testo divenuto p...,uri,https://fedlex.data.admin.ch/filestore/fedlex....
29806,uri,https://fedlex.data.admin.ch/eli/oc/2023/402,literal,Carta della Francofonia adottata ad Hanoi nel ...,literal,2023,literal,2005-11-23,literal,it,Trattato internazionale multilaterale,uri,https://fedlex.data.admin.ch/filestore/fedlex....
31362,uri,https://fedlex.data.admin.ch/eli/oc/2024/56,literal,Protocollo del 1996 alla Convenzione del 1972 ...,literal,2024,literal,2007-02-10,literal,it,Trattato internazionale multilaterale,uri,https://fedlex.data.admin.ch/filestore/fedlex....
36995,uri,https://fedlex.data.admin.ch/eli/oc/2023/454,literal,Accordo di consorzio ELIXIR del 26 giugno 2013...,literal,2023,literal,2014-01-12,literal,it,Trattato internazionale multilaterale,uri,https://fedlex.data.admin.ch/filestore/fedlex....
38669,uri,https://fedlex.data.admin.ch/eli/oc/2023/535,literal,Regolamento del Fondo nazionale svizzero del 2...,literal,2023,literal,2016-01-01,literal,it,Atto legislativo delle aziende e degli stabili...,uri,https://fedlex.data.admin.ch/filestore/fedlex....
...,...,...,...,...,...,...,...,...,...,...,...,...,...
46834,uri,https://fedlex.data.admin.ch/eli/oc/2024/688,literal,Ordinanza concernente gli effettivi massimi pe...,literal,2024,literal,2025-01-01,literal,it,Ordinanza del Consiglio federale,uri,https://fedlex.data.admin.ch/filestore/fedlex....
46835,uri,https://fedlex.data.admin.ch/eli/oc/2024/692,literal,Ordinanza concernente l’imposizione minima dei...,literal,2024,literal,2025-01-01,literal,it,Ordinanza del Consiglio federale,uri,https://fedlex.data.admin.ch/filestore/fedlex....
46836,uri,https://fedlex.data.admin.ch/eli/oc/2024/693,literal,Ordinanza sulla costituzione di una riserva di...,literal,2024,literal,2025-01-01,literal,it,Ordinanza del Consiglio federale,uri,https://fedlex.data.admin.ch/filestore/fedlex....
46837,uri,https://fedlex.data.admin.ch/eli/oc/2024/697,literal,Ordinanza sull’assicurazione malattie (OAMal),literal,2024,literal,2025-01-01,literal,it,Ordinanza del Consiglio federale,uri,https://fedlex.data.admin.ch/filestore/fedlex....


In [ ]:
# keep only the acts without a corresponding xml file
df = pd.read_csv("./raw_data/allActs.csv")
df_no_xml = df[df['fileUrl.value'].isna()]

# display the number of laws per publication year
df_countYear = df_no_xml
df_countYear['publicationDate.value'] = df_countYear['publicationDate.value'].str[:4]
print(df_countYear['publicationDate.value'].value_counts().sort_index(ascending=False))
# saved as acts_no_xml_year_agg.csv in the raw_data folder

print("------------------------------------------------------------------------------------")

# display the number of laws per publication year for the years 2022, 2023, 2024 and 2025
df_countYear = df_no_xml
df_countYear['publicationDate.value'] = df_countYear['publicationDate.value'].str[:4]
df_countYear = df_countYear[df_countYear['publicationDate.value'].isin(['2022', '2023', '2024', '2025'])]
print(df_countYear['publicationDate.value'].value_counts().sort_index(ascending=False))

print("------------------------------------------------------------------------------------")

# display the number of laws per type of document
df_countType = df_no_xml
df_countType['typeDoc.value'] = df_countType['typeDoc.value']
print(df_countType['typeDoc.value'].value_counts())

publicationDate.value
2025       1
2024     106
2023     150
2022     462
2021    1101
        ... 
1852       1
1851       2
1850       4
1849       2
1848       1
Name: count, Length: 177, dtype: int64
------------------------------------------------------------------------------------
publicationDate.value
2025      1
2024    106
2023    150
2022    462
Name: count, dtype: int64
------------------------------------------------------------------------------------
typeDoc.value
Ordinanza del Consiglio federale                                                 15561
Ordinanza di un dipartimento                                                      7504
Campo d'applicazione                                                              4416
Trattato internazionale bilaterale                                                4263
Ordinanza di un'ufficio                                                           3354
Trattato internazionale multilaterale                                            

/tmp/ipykernel_3461629/495628522.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_countYear['publicationDate.value'] = df_countYear['publicationDate.value'].str[:4]
/tmp/ipykernel_3461629/495628522.py:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_countYear['publicationDate.value'] = df_countYear['publicationDate.value'].str[:4]
/tmp/ipykernel_3461629/495628522.py:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexe

In [ ]:
df_xml = pd.read_csv("./raw_data/acts_xml.csv")
df_xml['publicationDate.value'] = pd.to_datetime(df_xml['publicationDate.value'])
df_xml['publicationDate.value'] = df_xml['publicationDate.value'].dt.strftime('%d.%m.%Y')

# download xmls and save them in folders by year with right name format
import os
import urllib.request
for index, row in df_xml.iterrows():
    year = row['publicationDate.value'][6:]
    if not os.path.exists(f"./LawsDocs/XMLs/{year}"):
        os.makedirs(f"./LawsDocs/XMLs/{year}")
    url = row['fileUrl.value']
    filename = row['act.value'].split('/')[-1] + "_" + row['publicationDate.value'][:5] + "_" + row['typeDoc.value']
    urllib.request.urlretrieve(url, f"./LawsDocs/XMLs/{year}/{filename}")
    break # remove in case of rerun, here only to stop the cycle 

In [ ]:
import xml.etree.ElementTree as ET
import os
import pandas as pd
from lxml import etree

def extract_href_links(xml_file):
    try:
        # Parse XML with error recovery
        parser = etree.XMLParser(recover=True)  # Allows parsing even if XML is broken
        tree = etree.parse(xml_file, parser)
        root = tree.getroot()

        # Extract all href links
        links = root.xpath("//@href")

        return links

    except etree.XMLSyntaxError as e:
        print(f"XML Syntax Error in file {xml_file}: {e}")
        return []

def get_hrefs_from_folder(folder):
    hrefs = []
    for file in os.listdir(folder):
        hrefs += extract_href_links(os.path.join(folder, file))
    return hrefs

hrefs = []
for year in os.listdir("./LawsDocs/XMLs"):
    year_folder = os.path.join("./LawsDocs/XMLs", year)
    if os.path.isdir(year_folder):
        hrefs += get_hrefs_from_folder(year_folder)

df_hrefs = pd.DataFrame(hrefs, columns=['href'])
df_hrefs.to_csv("./raw_data/hrefs.csv", index=False)

In [ ]:
df_hrefs = pd.read_csv("./raw_data/hrefs.csv")
df_hrefs = df_hrefs.dropna()
df_hrefs = df_hrefs[df_hrefs['href'].str.contains('https')]
df_hrefs = df_hrefs[df_hrefs['href'].str.contains('cc')]
df_hrefs.to_csv("./raw_data/hrefs_filtered.csv", index=False)

In [ ]:
df_hrefs = pd.read_csv("./raw_data/hrefs_filtered.csv")
# open web pages and check if they are still available
import requests

def check_url(url):
    try:
        response = requests.get(url)
        print(response.status_code)
        return response.status_code
    except requests.exceptions.RequestException as e:
        print(e)
        return e

status_codes = []
for index, row in df_hrefs.iterrows():
    status_code = check_url(row['href'])


Controllo quante delle leggi contenute in allActs fanno parte della RU e quante fanno parte della RS attraverso l'URL.

In [3]:
#filter the df by checking how many rows in the act.value column contain the word "cc" and how many the word "oc"
df = pd.read_csv("./raw_data/allActs.csv")
df_cc = df[df['act.value'].str.contains('cc')]
df_oc = df[df['act.value'].str.contains('oc')]
print(df_cc.shape[0])
print(df_oc.shape[0])

#check the publication year and entyintoforce year of the acts containing the word "cc"
print(df_cc['publicationDate.value'].value_counts())
print(df_cc['entryintoforce.value'].value_counts())

558
46128
publicationDate.value
2023-07-31    11
2021-11-19     7
2021-06-24     6
2022-03-24     6
2021-06-23     6
              ..
2022-10-13     1
2022-09-15     1
2022-09-08     1
2022-07-15     1
2022-08-23     1
Name: count, Length: 346, dtype: int64
entryintoforce.value
2024-01-01    49
2022-01-01    48
2025-01-01    44
2023-01-01    36
2021-07-01    19
              ..
2024-06-11     1
2024-08-14     1
2024-09-03     1
2025-02-01     1
2025-01-07     1
Name: count, Length: 206, dtype: int64


In [ ]:
import pandas as pd
df = pd.read_csv('./raw_data/allActs.csv')


,act.type,act.value,title.type,title.value,publicationDate.type,publicationDate.value,entryintoforce.type,entryintoforce.value,typeDoc.type,typeDoc.xml:lang,typeDoc.value,fileUrl.type,fileUrl.value
0,uri,https://fedlex.data.admin.ch/eli/oc/1/82_69_80,literal,Trattato d'estradizione tra la Svizzera e l'Im...,literal,1874-06-03,literal,1874-06-03,literal,it,Trattato internazionale bilaterale,NaN,NaN
1,uri,https://fedlex.data.admin.ch/eli/oc/18/654_588...,literal,Ordinanza concernente la consegna e il control...,literal,1901-04-19,literal,1901-04-19,literal,it,Ordinanza del Consiglio federale,NaN,NaN
2,uri,https://fedlex.data.admin.ch/eli/oc/18/865_796...,literal,Ordinanza del 9 novembre 1901 sull'esercizio d...,literal,1901-11-09,literal,1901-11-09,literal,it,Ordinanza del Consiglio federale,NaN,NaN
3,uri,https://fedlex.data.admin.ch/eli/oc/18/870_801...,literal,Ordinanza sull'esercizio della Fabbrica federa...,literal,1901-11-09,literal,1901-11-09,literal,it,Ordinanza del Consiglio federale,NaN,NaN
4,uri,https://fedlex.data.admin.ch/eli/oc/19/816_779...,literal,Decreto del Consiglio federale del 24 dicembre...,literal,1906-07-01,literal,1906-07-01,literal,it,Ordinanza del Consiglio federale,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
46850,uri,https://fedlex.data.admin.ch/eli/cc/2024/142,literal,Ordinanza della FINMA del 6 marzo 2024 sui ris...,literal,2024-04-05,literal,2025-01-01,literal,it,Atto legislativo delle aziende e degli stabili...,NaN,NaN
46851,uri,https://fedlex.data.admin.ch/eli/cc/2024/331,literal,Legge federale del 30 settembre 2022 sulla pro...,literal,2024-07-02,literal,2025-01-01,literal,it,Legge federale,NaN,NaN
46852,uri,https://fedlex.data.admin.ch/eli/cc/2024/493,literal,Ordinanza del 28 agosto 2024 sull’adeguamento ...,literal,2024-09-13,literal,2025-01-01,literal,it,Ordinanza del Consiglio federale,NaN,NaN
46853,uri,https://fedlex.data.admin.ch/eli/cc/2024/685,literal,Ordinanza del 13 novembre 2024 sulla cooperazi...,literal,2024-11-25,literal,2025-01-01,literal,it,Ordinanza del Consiglio federale,NaN,NaN


In [4]:
# filter the act.value keeping the last two elements of what we obtain from splitting the string by "/"
import pandas as pd
df = pd.read_csv('./raw_data/allActs.csv')
df['act.value'] = df['act.value'].str.split('/').str[-2:]
df


,act.type,act.value,title.type,title.value,publicationDate.type,publicationDate.value,entryintoforce.type,entryintoforce.value,typeDoc.type,typeDoc.xml:lang,typeDoc.value,fileUrl.type,fileUrl.value
0,uri,"[1, 82_69_80]",literal,Trattato d'estradizione tra la Svizzera e l'Im...,literal,1874-06-03,literal,1874-06-03,literal,it,Trattato internazionale bilaterale,NaN,NaN
1,uri,"[18, 654_588_629]",literal,Ordinanza concernente la consegna e il control...,literal,1901-04-19,literal,1901-04-19,literal,it,Ordinanza del Consiglio federale,NaN,NaN
2,uri,"[18, 865_796_844]",literal,Ordinanza del 9 novembre 1901 sull'esercizio d...,literal,1901-11-09,literal,1901-11-09,literal,it,Ordinanza del Consiglio federale,NaN,NaN
3,uri,"[18, 870_801_849]",literal,Ordinanza sull'esercizio della Fabbrica federa...,literal,1901-11-09,literal,1901-11-09,literal,it,Ordinanza del Consiglio federale,NaN,NaN
4,uri,"[19, 816_779_790]",literal,Decreto del Consiglio federale del 24 dicembre...,literal,1906-07-01,literal,1906-07-01,literal,it,Ordinanza del Consiglio federale,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
46850,uri,"[2024, 142]",literal,Ordinanza della FINMA del 6 marzo 2024 sui ris...,literal,2024-04-05,literal,2025-01-01,literal,it,Atto legislativo delle aziende e degli stabili...,NaN,NaN
46851,uri,"[2024, 331]",literal,Legge federale del 30 settembre 2022 sulla pro...,literal,2024-07-02,literal,2025-01-01,literal,it,Legge federale,NaN,NaN
46852,uri,"[2024, 493]",literal,Ordinanza del 28 agosto 2024 sull’adeguamento ...,literal,2024-09-13,literal,2025-01-01,literal,it,Ordinanza del Consiglio federale,NaN,NaN
46853,uri,"[2024, 685]",literal,Ordinanza del 13 novembre 2024 sulla cooperazi...,literal,2024-11-25,literal,2025-01-01,literal,it,Ordinanza del Consiglio federale,NaN,NaN


In [2]:
#duplicates = df[df.duplicated(keep=False)]
duplicate_counts = df['act.value'].value_counts()
duplicate_counts = duplicate_counts[duplicate_counts > 1]
duplicate_counts

act.value
[2021, 791]    3
[2024, 665]    3
[2024, 240]    3
[2021, 717]    3
[2023, 533]    3
              ..
[2023, 89]     2
[2024, 212]    2
[2024, 653]    2
[2023, 460]    2
[2023, 615]    2
Name: count, Length: 599, dtype: int64

In [32]:
dup_indices = pd.read_csv('./raw_data/duplicate_indices.csv')
# give me a list with all the indices
indices = dup_indices['Indices'].tolist()
# take all the values in the square brackets and put them in a list
indices_to = [int(i) for sublist in indices for i in sublist[1:-1].split(", ")]
indices.__len__(), indices_to.__len__()

(599, 1207)

In [36]:
#filter the df by the indices in the list
df_filtered = df.loc[indices_to]
# how many contanins "cc", "oc", "fga"
print(df_filtered['act.value'].str.contains('cc').sum(), "cc")
print(df_filtered['act.value'].str.contains('oc').sum(), "oc")
print(df_filtered['act.value'].str.contains('fga').sum(), "fga")


558 cc
599 oc
50 fga


Da questo risultato, siccome sappiamo che in 9 casi tutte e tre erano duplicate, possiamo ottenere che 599 - (50-9) = 558.
Ovvero tutti i "cc" sono duplicati sugli oc, quindi possono essere tutti rimossi comodamente.

In [39]:

df = pd.read_csv('./raw_data/allActs.csv')
# count "fga" in the act.value column
print(df['act.value'].str.contains('fga').sum())


169
